include/ccf/pal/snp_ioctl6.h

```diff
@@ -140,4 +198,50 @@ namespace ccf::pal::snp::ioctl6
       return {quote_bytes, quote_bytes + resp_wrapper.resp.report_size};
     }
   };
-}
+
+  class DerivedKey
+  {
+    DerivedKeyRespWrapper resp_wrapper = {};
+
+  public:
+    DerivedKey()
+    {
+      int fd = open(DEVICE, O_RDWR | O_CLOEXEC);
+      if (fd < 0)
+      {
+        throw std::logic_error(
+          fmt::format("Failed to open \"{}\" ({})", DEVICE, fd));
+      }
+
+      // This req by default mixes in HostData and the CPU VCEK
+      DerivedKeyReq req = {};
+      // We must also mix in the measurement
+      req.guest_field_select.measurement = 1;
+      GuestRequestDerivedKey payload = {
+        .req_data = &req, .resp_wrapper = &resp_wrapper, .exit_info = {0}};
+      int rc = ioctl(fd, SEV_SNP_GUEST_MSG_DERIVED_KEY, &payload);
+      if (rc < 0)
+      {
+        LOG_FAIL_FMT("IOCTL call failed: {}", strerror(errno));
+        LOG_FAIL_FMT(
+          "Exit info, fw_error: {} vmm_error: {}",
+          payload.exit_info.errors.fw,
+          payload.exit_info.errors.vmm);
+        throw std::logic_error(
+          "Failed to issue ioctl SEV_SNP_GUEST_MSG_DERIVED_KEY");
+      }
+      if ((*payload.resp_wrapper).resp.status != 0)
+      {
+        LOG_FAIL_FMT(
+          "SNP_GUEST_DERIVED_KEY failed: {}", resp_wrapper.resp.status);
+        throw std::logic_error(
+          "Failed to issue ioctl SEV_SNP_GUEST_MSG_DERIVED_KEY");
+      }
+    }
```

Comment:
Multiple questions on the storage format:

- Do we need to keep the entire `DerivedKeyRespWrapper` (including padding) around? Or can we just retain the actual key material?
- Do we need a class to hold this? I'd be tempted to just have a `std::vector<uint8_t> get_derived_key()` call, which would also make the errors nicer (avoid throwing in a constructor). If we think it's security-critical enough that we don't want to leave it in arbitrary/free-floating `std::vector`s fair enough, but then we should probably cleanse the memory on destruction as we do for other key-wrapping classes.

```suggestion
    }
    
    ~DerivedKey()
    {
      OPENSSL_cleanse(resp_wrapper.resp.data, sizeof(resp_wrapper.resp.data));
    }
```

Reply:
A vector by value is not great, partly because it's more difficult to make sure it's cleansed, and also because it means making an unnecessary copy. Having this be a type and having a way to get a read-only span is both safe and efficient.

Reply:
@eddyashton added the destructor. See discussion about padding below.